inference-engine/src/mkldnn_plugin/nodes/batch_to_space.cpp

```diff
@@ -3,219 +3,247 @@
-                        const size_t od_k = oc_k + od * OH_OW;
-                        for (; ih < IH; ih++) {
-                            const int64_t oh = ih * BSH + oh_add;
-                            if (oh < 0 || oh >= OH) {
-                                start += IW;
-                                if (start >= end)
-                                    break;
-                                continue;
-                            }
-                            const size_t oh_k = od_k + oh * OW;
-                            for (; iw < IW; iw++) {
-                                const int64_t ow = iw * BSW + ow_add;
-                                if (ow < 0 || ow >= OW) {
-                                    start++;
-                                    if (start >= end)
-                                        break;
-                                    continue;
+            splitter(workAmount, nthr, ithr, start, end);
+            std::vector<size_t> indxStart(2, 0);
+            std::vector<size_t> indxEnd(2, 0);
+            parallel_it_init(start, indxStart[0], inShape5D[0], indxStart[1], channels);
+            parallel_it_init((end - 1), indxEnd[0], inShape5D[0], indxEnd[1], channels);
+            for (size_t i0 = indxStart[0]; i0 < indxEnd[0] + 1; ++i0) {
+                int64_t bIdx = i0 / outShape5D[0];
+                const size_t srcIdx0 = i0 * inBatchStep;
+                const size_t dstIdx0 = (i0 - (bIdx * outShape5D[0])) * outBatchStep;
+                std::vector<int64_t> oAdd(5, 1);
+                std::vector<size_t> begin(5, 0);
+                std::vector<size_t> finish(5, 1);
+                oAdd[4] = bIdx % blockShapeIn[dimsSize - 1] - cropsBeginIn[dimsSize - 1];
+                bIdx /= blockShapeIn[dimsSize - 1];
+                oAdd[3] = bIdx % blockShapeIn[dimsSize - 2] - cropsBeginIn[dimsSize - 2];
+                bIdx /= blockShapeIn[dimsSize - 2];
+                oAdd[2] = dimsSize == 5 ? bIdx % blockShapeIn[2] - cropsBeginIn[2] : 0lu;
+                bIdx = dimsSize == 5 ? bIdx / blockShapeIn[2] : bIdx;
+                oAdd[1] = bIdx % blockShapeIn[1] - cropsBeginIn[1];
+                begin[1] = (blockShapeIn[1] - 1 - oAdd[1]) / blockShapeIn[1] / blockSize;
+                finish[1] = (outDims[1] - 1 - oAdd[1]) / blockShapeIn[1] / blockSize;
+                begin[2] = dimsSize == 5 ? (blockShapeIn[2] - 1 - oAdd[2]) / blockShapeIn[2] : 0lu;
+                finish[2] = dimsSize == 5 ? (outDims[2] - 1 - oAdd[2]) / blockShapeIn[2] : 0lu;
+                begin[3] = (blockShapeIn[dimsSize - 2] - 1 - oAdd[3]) / blockShapeIn[dimsSize - 2];
+                finish[3] = (outDims[dimsSize - 2] - 1 - oAdd[3]) / blockShapeIn[dimsSize - 2];
+                begin[4] = (blockShapeIn[dimsSize - 1] - 1 - oAdd[4]) / blockShapeIn[dimsSize - 1];
+                finish[4] = (outDims[dimsSize - 1] - 1 - oAdd[4]) / blockShapeIn[dimsSize - 1];
+                if (layout == NHWC || layout == NDHWC) {
```

Comment:
I still see some layout specific customization inside the loop. Lets discuss offline how to avoid it.

Reply:
Discussed offline: decision to keep current implementation.